### PR TITLE
qlat read and minor change to install file

### PR DESCRIPTION
**Pull request by jhrehanoaa**
qlat imports from yaml and cmd line
minor change to install file instructions
need to point this branch to fork not master

### CODE REVIEW

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 1]
```diff
@@ -167,10 +200,32 @@ def main():
...
+            run_parameters,
+        ) = nhd_io.read_custom_input(custom_input_file)
+
+        qlat_const = forcing_parameters.get("qlat_const", None)
+        qlat_input_file = forcing_parameters.get("qlat_input_file", None)
+        qlat_input_folder = forcing_parameters.get("qlat_input_folder", None)
+        qlat_file_pattern_filter = forcing_parameters.get(
+            "qlat_file_pattern_filter", None
+        )
+    else:
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:223]: I think we can leave this out of the if-block for this PR and just use the command-line inputs.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 2]
```diff
@@ -167,10 +200,32 @@ def main():
...
+        qlat_const = float(args.qlat_const)
+        qlat_input_folder = args.qlat_input_folder
+        qlat_input_file = args.qlat_input_file
+        qlat_file_pattern_filter = args.qlat_file_pattern_filter
+    # print(forcing_parameters,qlat_const)
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:228]: Down below here, where we have this (:270-273 in the modified source)
```
   if args.ql:
        qlats = nhd_io.read_qlat(args.ql)
    else:
        qlats = constant_qlats(param_df, nts, 10.0)
```
we need to replace it with something like this (from src/python_routing_v01/compute_nhd_routing_SingleSeg.py:1589-1621)
```
    # STEP 5: Read (or set) QLateral Inputs
    if showtiming:
        start_time = time.time()
    if verbose:
        print("creating qlateral array ...")

    # initialize qlateral dict
    qlateral = {}

    if qlat_input_folder:
        qlat_files = glob.glob(qlat_input_folder + qlat_file_pattern_filter)
        qlat_df = nio.get_ql_from_wrf_hydro(
            qlat_files=qlat_files,
            index_col=qlat_file_index_col,
            value_col=qlat_file_value_col,
        )

    elif qlat_input_file:
        qlat_df = nio.get_ql_from_csv(qlat_input_file)

    else:
        qlat_df = pd.DataFrame(
            qlat_const, index=connections.keys(), columns=range(nts), dtype="float32"
        )

    for index, row in qlat_df.iterrows():
        qlateral[index] = row.tolist()

    if verbose:
        print("qlateral array complete")
    if showtiming:
        print("... in %s seconds." % (time.time() - start_time))
        start_time = time.time()
```

**Review by jameshalgren — CHANGES_REQUESTED** [review]
A couple things to address.

File: src/python_framework_v02/troute/nhd_io.py [hunk 3]
```diff
@@ -42,7 +42,7 @@ def read_mask(path, layer_string=None):
 def read_custom_input(custom_input_file):
     if custom_input_file[-4:] == "yaml":
         with open(custom_input_file) as custom_file:
-            data = yaml.safe_load(custom_file)
+            data = yaml.load(custom_file, Loader=yaml.SafeLoader)
```
Comment (jameshalgren) [inline src/python_framework_v02/troute/nhd_io.py:45]: Make sure to rebase to the latest master -- I think we have this in there already.

**Review by jameshalgren — COMMENTED** [review]
Remember to:
```
git fetch upstream
git rebase upstream/master
```

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py (hunk 1 above)
Comment (jhrehanoaa) [reply src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:223]: Won't the end result be using a custom input file anyway? I figured right now it loads in and we can take what we want from it and disregard the rest. Wouldn't command-line be going backward?

File: src/python_framework_v02/troute/nhd_io.py (hunk 3 above)
Comment (jhrehanoaa) [reply src/python_framework_v02/troute/nhd_io.py:45]: I have a fresh install. The original safe_load is in already, but after reviewing it online from a link it was displaying, the second choice is the more proper use.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 4]
```diff
@@ -212,10 +270,41 @@ def main():
...
-    if args.ql:
-        qlats = nhd_io.read_qlat(args.ql)
+    # STEP 5: Read (or set) QLateral Inputs
+    if showtiming:
+        start_time = time.time()
+    if verbose:
+        print("creating qlateral array ...")
+
+    # initialize qlateral dict
+    qlateral = {}
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:280]: I think we can strike the qlateral dictionary... it must be leftover from prior versions.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 5]
```diff
@@ -260,15 +316,53 @@ def main():
...
 
+    # STEP 5: Read (or set) QLateral Inputs
+    if showtiming:
+        start_time = time.time()
+    if verbose:
+        print("creating qlateral array ...")
+
+    # initialize qlateral dict
+    qlateral = {}
+
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:327]: the `qlateral` object is no longer used...

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 6]
```diff
@@ -260,15 +316,53 @@ def main():
...
+    if qlat_input_folder:
+        qlat_files = glob.glob(qlat_input_folder + qlat_file_pattern_filter)
+        qlat_df = nhd_io.get_ql_from_wrf_hydro(
+            qlat_files=qlat_files,
+            index_col=qlat_file_index_col,
+            value_col=qlat_file_value_col,
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:333]: We'll need to add CLI args for these two values. This is a bugfix issue to propagate back to V01, also.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 7]
```diff
@@ -260,15 +316,53 @@ def main():
...
+        )
+
+    elif qlat_input_file:
+        qlat_df = nhd_io.get_ql_from_csv(qlat_input_file)
+
+    else:
+        qlat_df = pd.DataFrame(
+            qlat_const, index=connections.keys(), columns=range(nts), dtype="float32"
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:341]: The qlf and qlw options produce dataframes that will need to be checked and trimmed/filled-with-zeros so they are the same size as this constant input qlateral that we've defined as the most basic input option.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 8]
```diff
@@ -260,15 +316,53 @@ def main():
...
+        )
+
+    qlats = qlat_df
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:344]: Please add a #TODO: Create pulse or other useful synthetic qlateral input tests...

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py (hunk 7 above)
Comment (jhrehanoaa) [reply src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:341]: [28 rows x 144 columns] - working constant qlat size 
[374 rows x 24 columns] - broken command size

I'm going to go ahead and trim the rows and add a bunch of zeros to the columns then.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 9]
```diff
@@ -260,15 +338,61 @@ def main():
...
+    
     cpu_pool = args.cpu_pool
     compute_method = args.compute_method
 
     if compute_method == "standard cython compute network":
         compute_func = mc_reach.compute_network
     else:
         compute_func = mc_reach.compute_network
-
+
```
Comment (jameshalgren) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:395]: Run black on this and it should be ready.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py (hunk 7 above)
Comment (jameshalgren) [reply src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py]: This sizing issue seems to be the last remaining question. (Saw that the file was blackened ...)

Comment (jameshalgren) [reply src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py]: > trim the rows
The rows represent the segment ID (i.e., `connections`) and there are several dataframes which use that table and could be the source of an index to filter the qlat dataframe... 

> add a bunch of zeros
That's probably the right first step. The 24 hours in the current dataset will need to be subdivided into 300s timesteps, but let's get a test with zero padding to work first.

Comment (jameshalgren) [reply src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py]: @jhrehanoaa 
Please add an issue to implement time step adjustment for qlaterals in v02. Please also add an issue to validate the v02 computation against the Pocono_TEST2 test case.

**Review by jhrehanoaa — COMMENTED** [review]
-

**Review by jameshalgren — APPROVED** [review]
Please squash and merge after creating the noted issues: 
Add qlateral timestep adjustment capability to v02
Validate v02 computation against Pocono_TEST2 test results
